lib/ansible/modules/packaging/os/satellite_subscription.py

```diff
@@ -0,0 +1,442 @@
+    Your hosts or hypervisors should be in the Satellite already. The module will run subscription
+    related tasks (assign, remove, autoattach or replace with hypervisor subscriptions) for the
+    hosts through the Satellite API.
+version_added: "2.8"
+requirements:
+  - requests
+author:
+  - Luc Stroobant (@stroobl)
+options:
+  satellite_hostname:
+    description:
+      - The name or IP address of the Red Hat Satellite server.
+    required: true
+  username:
+    description:
+      - Satellite server username
+    required: true
+  password:
+    description:
+      - Satellite server password
+    required: true
+  organization:
+    description:
+      - Satellite organization ID to work on
+    required: true
+  state:
+    description:
+      - "present: assign a subscription of subscription_type to host"
+      - "absent: remove subscription of subscription_type from host"
+      - "autoattach: run subscription autoattach for host (autoattach will be enabled if necessary)"
+      - "disable_autoattach: disable subscription autoattach for host"
+      - "vdcguests: remove all Red Hat subscriptions from a hypervisor's guests and run autoattach
+         to replace them with virtual guest subscriptions. Run this against a virt-who hypervisor to
+         remove all other subscriptions assigned to the guests. Useful if you configured a VDC
+         subscription on a hypervisor that didn't have it before."
+      - "optimize: use to replace subscriptions on physical servers with a higher multiplier with subscriptions with a lower multiplier."
+    choices: [present, absent, autoattach, disable_autoattach, vdcguests, optimize]
+    default: autoattach
+  content_host:
+    description:
+      - List of content hosts to work with
+      - Can also be a hostname search pattern accepted by the satellite.
+  subscription:
+    description:
+      - List of subscriptions to assign or remove
+      - Required when action is assign or remove
+      - "This field is used for the search for subscriptions, it can contain (part of) the subscription
+         name, eg 'Red Hat Enterprise Linux Server Entry Level, Self-support' or another search query
+         as used in the satellite, eg: 'product_id=RH00005S'"
+  unique:
+    description:
+     - Whether to remove all subscriptions to this host.
+     - When set to C(yes) remove all other subscriptions attached to this host.
+    type: bool
+    default: no
+  verify_ssl:
+    description:
+      - Check ssl certificates?
+    type: bool
+    default: True
+"""
+
+EXAMPLES = """
+- name: Assign subscriptions, remove all existing subscriptions (run against localhost)
+  satellite_subscription:
+    satellite_hostname: "{{ satellite_fqdn }}"
+    username: "{{ satellite_admin_user }}"
+    password: "{{ satellite_admin_pass }}"
+    state: present
+    content_host: "{{ groups.rhel_servers }}"
+    subscription:
+      - product_id=RH000XX
```

Comment:
Would this not be better a list of dicts ?

Also, is the number of possible keys in that dict limited to e.g. `product_id` and `name`, or is it free-form ?
Because you can actually enforce certain values in dictionary lists from the argspec.

Reply:
It's not necessary a key-value. In the example on line 123 you have:
```
subscription:
 - "Red Hat Enterprise Linux Server Entry Level, Self-support"
```
as far as I can test, the api is not 100% consistent there, it doesn't accept a search key if you search for a name. (so name=Red Hat Enterprise Linux Server Entry Level, Self-support doesn't work).